File: pow2_native.hs
{-# LANGUAGE RankNTypes #-}

import System.Environment

data Tree = Bin Tree Tree | Tip

size :: Tree -> Int
size Tip       = 1
size (Bin a b) = size a + size b

pow2 :: Int -> Tree
pow2 0 = Tip
pow2 n = Bin (pow2 (n - 1)) (pow2 (n - 1))

main :: IO()
main = do
  args <- getArgs
  let depth = if length args < 1 then 0 else read (head args)
  print (size (pow2 depth))

File: pow2_native.js
const Bin = (a, b) => ({ctor:"Bin", a, b});
const Tip = {ctor:"Tip"};

const size = (tree) => {
  switch (tree.ctor) {
    case "Tip": return 1;
    case "Bin": return size(tree.a) + size(tree.b);
  }
};

const pow2 = (n) => {
  if (n === 0) {
    return Tip;
  } else {
    return Bin(pow2(n - 1), pow2(n - 1));
  };
};

var depth = Number(process.argv[2]||"0");
console.log(size(pow2(depth)));
